readme
What is huibase ?
==========
huibase a basic static c++ library created by TomHui at 2011. 

Requirement:
==========
1> run function in hcapp, set to return hret;
2> name of HCHapp;

Conclusion:
==========
1> All classes needs to declare default, copy, move construction functions and assignment functions
as far as possible;
2> Check the virtual property before you inherit a class;


Modify History
==========
Add network address and socket component. TomHui@Sun Sep 17 CST 2017.
HCAddr is basic class for all address classes, and HCSocket is base
class for all socket classes.

Add basic pthread component. TomHui@Thu Sep 21 CST 2017.
Include: pthread_attr_t, pthread_t, pthread_mutex_t, pthread_mutex_attr_t, pthread_cond_t, pthread_cond_attr_t;

Fix the Rtrim function. TomHui@Mon Sep 25 CST 2017.

Fix the HCSocket::GetLocalIp function. TomHui@Fri Sep 29 CST 2017.
//TODO local ip; proc number in 'ps aux'

Oct 09 2017 Add HCMq. TomHui@Mon Oct 09 CST 2017.
shareMemory:
node_header |nodex_id(4),name(HNAME),len(4),status(1)		|
------------|-----------------------------------------------|
block_header|block_id(4),ctime(4),mtime(4),len(4),char[MAX_SHM_BLOCK_LEN],status	|
			|
------------|

shminfo: share memory to save the information of process.


task category: 
	1> task unit: config map;
	2> command: 
	3> server job:


==========
Add OS name to makefile. Edit code at inc.mk. TomHui@Sun Oct 10 CST 2017.

==========
Map object add the default value. TomHui@Wed Oct 18 CST 2017.

==========
Add CDic, CFileName, CFile

==========
Add Stop to happ. TomHui@Sun Nov 12 1800 CST 2017.

==========
Add HCMutex holder. TomHui@Sun Jan 3 CST 2018.

==========
Remove keep socket alive in tcp socket class. TomHui@Tue Jun 19 CST 2018.

==========
Version Code: 1.1.2
Handle redis's protocal exception.

history.md
# Modify History

## v4S1E1
* VERSION CREATE DATE: Fri Dec 20 11:07 CST 2019
* VERSION AUTHOR: Tom Hui
* VERSION COMMIT: 
	> thread_key_t, CThreadKey;


## v4S1
* VERSION CREATE DATE: Wed Dec 18 17:52 CST 2019
* VERSION AUTHOR: Tom Hui
* VERSION COMMIT: 
	> Update makefile;


## v4
* VERSION CREATE DATE: Mon Dec 16 20:54 CST 2019
* VERSION AUTHOR: Tom Hui
* VERSION COMMIT: 
	> Update md format;
	> Add makerel.sh;
	> Make HApps Stop() virtual;



## V1S14
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S14
* VERSION CREATE DATE: Thu Aug 16 15:00 CST 2018
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> hctime c++ template strtime.

* PLAN:



## V1S12
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S12
* VERSION CREATE DATE: Sat Jul 28 1500 CST 2018
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> update redis lib;
* PLAN:
		> check redis's memory;



## V1S11
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S11
* VERSION CREATE DATE: Sat Jul 28 1500 CST 2018
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> rebuild huidb;
* PLAN:



## V1S10
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S10
* VERSION CREATE DATE: Thu Jul 26 1500 CST 2018
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> Memcached increment/decrement.
* PLAN:
		> Add int64_t, uint64_t to base.



## V1S7
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S7
* VERSION CREATE DATE: Sat Jun 23 1500 CST 2018
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> Memcached buffer length to 20480




## V1S6
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S6
* VERSION CREATE DATE: Sat Jun 23 1500 CST 2018
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> Convert k-v string with HCParam;




## V1S5
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S5
* VERSION CREATE DATE: Sat Jun 23 0922 CST 2018
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> Fix the bug that hlog shift day, it cann't change the size file name;




## V1S5
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S5
* VERSION CREATE DATE: Thu Nov 08 1922 CST 2018
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> Update huibase redis client library;




## V1S3E1
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S3E1
* VERSION CREATE DATE: Sat Feb 16 21:07:03 CST 2019
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> Update mysql lib;
		> Add an error code for mysqllib;




## V1S5E1 		
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S5E1
* VERSION CREATE DATE: Tue Mar 05 16:32 CST 2019
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> Update log flags;
		> HBOOL to bool;




## V1S8E1
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S8E1
* VERSION CREATE DATE: Web Mar 27 23:19 CST 2019
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> add Count for HCStr;



## V1S8E3
* PROJECT NAME: HUIBASE
* VERSION CODE: V1S8E3
* VERSION CREATE DATE: Web Apr 3 17:19 CST 2019
* VERSION AUTHOR: Tom Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> print system log in the thread pool;




## V2S1E1
* PROJECT NAME: HUIBASE
* VERSION CODE: V2S1E1
* VERSION CREATE DATE: Mon Apr 8 17:19 CST 2019
* VERSION AUTHOR: Joseph Hui 8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> throw std::redis_exception in the libhuinosql;




## V2S1E1
* PROJECT NAME: HUIBASE
* VERSION CODE: V2S1E1
* VERSION CREATE DATE: Tue Apr 9 15:31 CST 2019
* VERSION AUTHOR: Joseph Hui +8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> pool write system log;




## V2S1E3
* PROJECT NAME: HUIBASE
* VERSION CODE: V2S1E3
* VERSION CREATE DATE: Sat Apr 13 15:31 CST 2019
* VERSION AUTHOR: Joseph Hui +8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> Memory leak detected in hlog; set the m_fd as HINVA_FD in init;




## V2S3
* PROJECT NAME: HUIBASE
* VERSION CODE: V2S3
* VERSION CREATE DATE: Web Ja6 8 15:31 CST 2019
* VERSION AUTHOR: Joseph Hui +8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> change hlog, make sure it works in multiple-thread environment;
		> add repo-map;
		> happ add get config value interface;



## v8s1
* PROJECT NAME: HUIBASE
* VERSION CODE: V8S1
* VERSION CREATE DATE: Sat Aug 22 15:39 CST 2020
* VERSION AUTHOR: Joseph Hui +8613760232170. tomhui1009@yahoo.com
* VERSION COMMENT:
		> compile huibase with g++ 9.3
		> add repo-map;
		> happ add get config value interface;



src/hstr.h
/**************************=cpphd=******************************************
 * 
 * @ File Name: hstr.h
 * 
 * @ Create Author: Tom Hui
 * 
 * @ Create Date: Fri Dec 23 14:17 2016
 * 
 * @ Brief:
 * 	This file declare all functions to process strings.
 * 
 ****************************************************************************/


#ifndef _H_STR_H_
#define _H_STR_H_

#include "huibase.h"

namespace HUIBASE{

/*************************=cldes=**********************************
 @ Class Name:  HCStr
 @ Create Author: Tom Hui
 @ Create Date: Tue Dec 20 10:19 2016
 @ Description:
	manage string function and extend string.
 ******************************************************************/
    class HCStr : public HSTR{
    public:
		typedef HSTR base;
		typedef base::value_type value_type;
		typedef base::size_type size_type;
		typedef base::traits_type traits_type;
		typedef base::allocator_type allocator_type;

    public:
	HCStr()
	{}

    HCStr(const HCStr & str) : base(str)
	{}

	HCStr(const base& str) : base(str)
	{}

	HCStr(const value_type * sz) : base(sz)
	{}

	HCStr(const value_type * c, size_type n, const allocator_type & al = allocator_type())
	    : base(c, n, al)
	{}

	HCStr(const base & b, size_type pos, size_type count = base::npos )
	    : base(b, pos, count)
	{}

	HCStr(size_type c, value_type v)
	    :base(c, v)
	{}
	
    public:
	HCStr& operator= (const value_type * vt) {

	    assign(vt);

	    return *this;
	    
	}

	HCStr& operator= (const HCStr& str) {

	    assign(str);

	    return *this;

	}

    public:
	HINT split(HCSTRR key, HVSTRSR res) const {

	     return Split(*this, key, res);

	}
	
	HCStr & upper() {

	    Upper(*this);

	    return *this;

	}

	HCStr& lower() {

	    Lower(*this);

	    return *this;

	}

	HCStr ltrim(HCPSZ sz = NULL) {

	    Ltrim(*this, sz);

	    return *this;

	}
	

	HCStr& rtrim (HCPSZ sz = NULL) {

	    Rtrim(*this, sz);

	    return *this;

	}

	
	HCStr& trim(HCPSZ sz = NULL) {

	    Trim(*this, sz);

	    return *this;

	}


	HBOOL isDigit() const {

	    return IsDigit(*this);

	}

	HCStr& right(HSIZE len) {

	    Right(*this, len);

	    return *this;

	}

	HCStr& left(HSIZE len) {

	    Left(*this, len);

	    return *this;

	}

	HCStr& replace(HCSTRR srckey, HCSTRR dstKey) {

	    Replace(*this, srckey, dstKey);

	    return *this;

	}

	HCStr& replaces(HCSTRR srcKey, HCSTRR dstKey) {

	    Replaces(*this, srcKey, dstKey);

	    return *this;

	}


	HINT CountTime (HCSTRR key) {

		return CountTime(*this, key);

	}

	
	HCStr& remove(HCSTRR srcKey) {

	    Remove(*this, srcKey);

	    return *this;

	}

	HCStr& removes(HCSTRR srcKey) {

	    Removes(*this, srcKey);

	    return *this;

	}

	HCStr& appendFormat(const value_type* fmt, ... );
	
    public:
	
	/*****************************=memfu=*************************************
	 @ Member Function Name: Split
	 @ Create Author: Tom Hui
	 @ Create Date: Tue Dec 20 11:09 2016
	 @ Parameters:
	       1> str: source string function will operate it;
               2> key: split key;
               3> res: return buffer;
	 @ Return Value:
   	      success 0, otherwide a herror will be signed
	*************************************************************************/ 
	static HINT Split(HCSTRR str, HCSTRR key, HVSTRSR res );

	static HCSTR s_tr_str;
	/*****************************=memfu=*************************************
	 @ Member Function Name: Ltrim
	 @ Create Author: Tom Hui
	 @ Create Date: Tue Dec 20 11:28 2016
	 @ Parameters: 
	      1> str: source string
              2> sz: 
	 @ Return Value:
   	      success 0, otherwide a herror will be signed.
	*************************************************************************/	
	static HINT Ltrim( HSTRR str, HCPSZ sz  = NULL ) ;


	/*****************************=memfu=*************************************
	 @ Member Function Name: Rtrim
	 @ Create Author: Tom Hui
	 @ Create Date: Tue Dec 20 11:42 2016
	 @ Parameters:
	     1> str: source string
	     2> sz: 
	 @ Return Value:
   	      success 0, otherwide a herror will be signed   	
	*************************************************************************/	
	static HINT Rtrim( HSTRR str, HCPSZ sz  = NULL  );

	static HINT Trim( HSTRR str, HCPSZ sz  = NULL  );

	static HINT Upper (HSTRR str);

	static HINT Lower(HSTRR str);

	static HBOOL IsDigit(HCSTRR str);

	static HBOOL IsDigitAndDos (HCSTRR str);

    static HBOOL IsDigitChar (HCSTRR str);

	static HINT Right(HSTRR str, HSIZE len);

	static HINT Left(HSTRR str, HSIZE len);

/*****************************=memfu=*************************************
 * @ Create Author: Tom Hui
 * @ Create Date: Tue Dec 27 10:53 2016
 * @ brief:
 *	replace a string speacified by srcKey in source string with dst string.
 * @ Param:
 *	1> src: source string;
 *	2> srckey: src key;
 *      3> dstKey: dst key;
 * @ Return Value:
 *   	success return OK.
 *************************************************************************/	
	static HINT Replace (HSTRR src, HCSTRR srcKey, HCSTRR dstKey);

/*****************************=memfu=*************************************
 * @ Create Author: Tom Hui
 * @ Create Date: Tue Dec 27 10:56 2016
 * @ brief:
 *	replace all strings in source string with dst string
 * @ Param:
 *	1> src: 
 *	2> srcKey:
 *      3> dstKey:
 * @ Return Value:
 *   	success return OK.
 *************************************************************************/	
	static HINT Replaces (HSTRR src, HCSTRR srcKey, HCSTRR dstkey);

	static HINT CountTime (HCSTRR src, HCSTRR key);

/*****************************=memfu=*************************************
 * @ Create Author: Tom Hui
 * @ Create Date: Tue Dec 27 14:21 2016
 * @ brief:
 *	move a key string specitied by strKey from src.
 * @ Param:
 *	1> src: source string
 *	2> strKey: key string
 * @ Return Value:
 *   	success return OK
 *************************************************************************/	
	static HINT Remove(HSTRR src, HCSTRR strKey);

	static HINT Removes(HSTRR src, HCSTRR strKey);

	static HSTR V2Str (HCVSTRSR strs, HSTR l = "|");

	static HCStr Format(const value_type* fmt, ...);

	static HCStr vFormat(const value_type* fmt, va_list ap);

	static HSTR WebStr (HCSTRR str, HUINT n);

	static HSTR WebStr (HDOUBLE d, HUINT n);

	static HINT stoi (HCSTRR str) {
	    return std::stoi(str);
	}
	
	static HINT stoi (HCPSZ sz) {
	    return std::stoi(HSTR(sz));
	}

    static HLONG stol (HCSTRR str);

	static HDOUBLE stod(HCSTRR str);

	static HN64 ston64(HCSTRR str);

    static HLLINT stoll(HCSTRR str);

    static HSTR lltos (HLLINT ll);

	static HSTR itos (HINT i);

	static HSTR dtos(HDOUBLE d);


    };

}

#endif


